streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java

```diff
@@ -0,0 +1,297 @@
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Properties;
+import java.util.Set;
+import java.util.TreeSet;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.core.IsEqual.equalTo;
+
+public class QueryableStateIntegrationTest {
+
+    @ClassRule
+    public static final EmbeddedSingleNodeKafkaCluster CLUSTER =
+        new EmbeddedSingleNodeKafkaCluster();
+    private static final String STREAM_ONE = "stream-one";
+    private static final String STREAM_TWO = "stream-two";
+    private static final String OUTPUT_TOPIC = "output";
+
+    private Properties streamsConfiguration;
+    private KStreamBuilder builder;
+    private KafkaStreams kafkaStreams;
+    private Comparator<KeyValue<String, String>> stringComparator;
+    private Comparator<KeyValue<String, Long>> stringLongComparator;
+
+    @BeforeClass
+    public static void createTopics() {
+        CLUSTER.createTopic(STREAM_ONE);
+        CLUSTER.createTopic(STREAM_TWO);
+        CLUSTER.createTopic(OUTPUT_TOPIC);
+    }
+
+    @Before
+    public void before() throws IOException {
+        builder = new KStreamBuilder();
+        streamsConfiguration = new Properties();
+        final String applicationId = "queryable-state";
+
+        streamsConfiguration.put(StreamsConfig.APPLICATION_ID_CONFIG, applicationId);
+        streamsConfiguration
+            .put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers());
+        streamsConfiguration.put(StreamsConfig.ZOOKEEPER_CONNECT_CONFIG, CLUSTER.zKConnectString());
+        streamsConfiguration.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        streamsConfiguration.put(StreamsConfig.STATE_DIR_CONFIG,
+                                 TestUtils.tempDirectory(new File("/tmp").toPath(), "qs-test")
+                                     .getPath());
+        streamsConfiguration.put(StreamsConfig.KEY_SERDE_CLASS_CONFIG, Serdes.String().getClass());
+        streamsConfiguration
+            .put(StreamsConfig.VALUE_SERDE_CLASS_CONFIG, Serdes.String().getClass());
+        stringComparator = new Comparator<KeyValue<String, String>>() {
+
+            @Override
+            public int compare(final KeyValue<String, String> o1,
+                               final KeyValue<String, String> o2) {
+                return o1.key.compareTo(o2.key);
+            }
+        };
+        stringLongComparator = new Comparator<KeyValue<String, Long>>() {
+
+            @Override
+            public int compare(final KeyValue<String, Long> o1,
+                               final KeyValue<String, Long> o2) {
+                return o1.key.compareTo(o2.key);
+            }
+        };
+    }
+
+    @After
+    public void shutdown() throws IOException {
+        if (kafkaStreams != null) {
+            kafkaStreams.close();
+        }
+        IntegrationTestUtils.purgeLocalStreamsState(streamsConfiguration);
```

Comment:
Ah, I see that the state directory is being purged here.  But I think this @After-approach may not work if the test crashes or the JVM terminates before `shutdown` is being called, right? (e.g. due to Jenkins woes) I think it would be safer to purge the state dir prior to the run.